File: test/fixtures/codex-account-handoff-pty.py
"""Private synthetic account-handoff driver; no ordinary prompt or real tab.

stdin/stdout carry a bounded fixture RPC. Only the already approved continuity
thread and two retained sign-ins are allowed. The persistent shell owns this
one private PTY. Neither another process nor a real Terminal window is targeted.
"""
import hashlib
import atexit
import json
import os
from pathlib import Path
import re
import shlex
import signal
import subprocess
import sys
import time
import pexpect
import pyte

base = Path.home() / 'Library/Application Support/ClawDad/Accounts/verification-2026-09-15'
fixture = base / 'thread-continuity-1'
if len(sys.argv) != 2 or not re.fullmatch(r'handoff-pty-[a-z0-9-]+', sys.argv[1]):
    raise SystemExit('Supply a new handoff-pty-* receipt name.')
root = fixture / sys.argv[1]
root.mkdir(mode=0o700)
original = json.loads((fixture / 'evidence.json').read_text())
thread = original['sourceThreadId']
project = str(fixture / 'project')
if original['modelTurns'] != 1 or original['project'] != project or thread != '01a0a848-cb86-7033-ae6f-ce006f5b51bb':
    raise SystemExit('Only the exact prior synthetic continuity fixture is allowed.')
logs = list((fixture / 'sessions').rglob('rollout-*-' + thread + '.jsonl'))
if len(logs) != 1:
    raise SystemExit('The fixture must have one exact persisted history.')
history = logs[0]
encoder = json.JSONEncoder(ensure_ascii=False, separators=(',', ':'))
def digest(text):
    return hashlib.sha256(text.encode()).hexdigest()
def write(name, value):
    file = root / name
    file.write_text(value)
    file.chmod(0o600)
def accepted():
    selected = []
    for line in history.read_text().splitlines():
        row = json.loads(line)
        payload = row.get('payload', {})
        if (row.get('type') == 'response_item' and payload.get('role') in ('user', 'assistant')) or (row.get('type') == 'event_msg' and payload.get('type') in ('user_message', 'task_started', 'task_complete', 'turn_aborted')):
            selected.append(row)
    return digest(encoder.encode(selected))
baseline = accepted()
env = {k: os.environ[k] for k in ('PATH', 'HOME', 'USER', 'LOGNAME', 'TMPDIR', 'LANG') if k in os.environ}
env.update({'TERM': 'xterm-256color', 'ZDOTDIR': str(root), 'HISTFILE': str(root / 'shell-history')})
shell = pexpect.spawn('/bin/zsh', ['-f'], cwd=project, env=env, encoding='utf-8', timeout=25, dimensions=(48, 180))
atexit.register(lambda: shell.close(force=True) if shell.isalive() else None)
write('fixture-shell.json', encoder.encode({'pid': shell.pid, 'parent': os.getpid(), 'project': project}))
# child_fd is the master; obtain the slave from the exact spawned shell PID.
slave = subprocess.check_output(['/bin/ps', '-p', str(shell.pid), '-o', 'tty='], text=True).strip()
if not re.fullmatch(r'ttys[0-9]+', slave):
    shell.close(force=True)
    raise SystemExit('The private shell did not expose one TTY.')
tty = '/dev/' + slave
screen = pyte.Screen(180, 48)
stream = pyte.Stream(screen)
shell.send("PROMPT='CLAWDAD_FIXTURE> '; RPROMPT=''; unset HISTFILE\r")
current = None
known = ''
status = None
receipts = {}
serial = 0
raw = ''

def receive():
    global raw
    try:
        text = shell.read_nonblocking(size=65536, timeout=0.15)
    except pexpect.TIMEOUT:
        return
    raw += text
    if len(raw) > 4 * 1024 * 1024:
        raise RuntimeError('Bounded fixture terminal output exceeded.')
    stream.feed(text)
    if '\x1b[6n' in text:
        shell.send('\x1b[%d;%dR' % (screen.cursor.y + 1, screen.cursor.x + 1))

def visible():
    return '\n'.join(screen.display)

def wait(predicate, seconds=12):
    deadline = time.monotonic() + seconds
    while time.monotonic() < deadline:
        receive()
        if predicate(visible()):
            return
    raise RuntimeError('The private fixture did not reach the expected observed state.')

def owner():
    rows = subprocess.check_output(['/bin/ps', '-t', slave, '-o', 'pid=,pgid=,tpgid=,stat=,lstart=,comm='], text=True)
    matches = []
    for line in rows.splitlines():
        columns = line.split(maxsplit=9)
        if len(columns) == 10 and columns[1] == columns[2] and Path(columns[9]).name == 'codex':
            matches.append(columns)
    if len(matches) != 1:
        return None
    row = matches[0]
    files = subprocess.check_output(['/usr/sbin/lsof', '-a', '-p', row[0], '-Fn'], text=True)
    if 'n' + str(history) not in files.splitlines():
        raise RuntimeError('The foreground process does not hold the exact synthetic history.')
    return {'pid': int(row[0]), 'processIdentity': 'codex-process-' + digest('|'.join(row[:3] + row[4:]))}

def composer(text):
    rows = text.splitlines()
    starts = [i for i, line in enumerate(rows) if line.strip().startswith('›')]
    if not starts:
        return None
    i = starts[-1]
    return rows[i].strip()[1:].strip()

def verify_draft():
    value = composer(visible())
    if not known:
        return value == '' or value == 'Ask Codex to do anything'
    if '\n' not in known and len(known) < 140:
        return value == known
    return value == '[Pasted Content %d chars]' % len(known)

def capture_status(profile):
    global status
    if known or not verify_draft():
        raise RuntimeError('Status needs the exact empty disposable composer.')
    before = owner()
    if not before:
        raise RuntimeError('No exact disposable agent owner.')
    shell.send('/status')
    wait(lambda text: composer(text) == '/status')
    time.sleep(0.25)
    shell.send('\r')
    email = 'codyshanemitchell@gmail.com' if profile == 'cody' else 'playinthesunwithme@gmail.com'
    wait(lambda text: email in text and 'Account:' in text and thread in text)
    # Let the compositor finish and observe the prompt after the complete panel.
    wait(lambda text: composer(text) in ('', 'Ask Codex to do anything'))
    if owner() != before or accepted() != baseline:
        raise RuntimeError('The owner or accepted work changed during local status.')
    text = visible()
    if 'gpt-6-astra (reasoning low, summaries auto)' not in text or 'Read Only (never)' not in text:
        raise RuntimeError('The expected fixture model and permissions were not shown.')
    status = {'profile': profile, 'email': email, 'model': 'gpt-6-astra', 'reasoningEffort': 'low', 'permissions': 'read-only/never', **before}
    write('status-%d-%s.txt' % (serial, profile), text)

def observe():
    if accepted() != baseline:
        raise RuntimeError('The fixture accepted turn history changed. No further action is permitted.')
    value = owner()
    common = {'tty': tty, 'tabLifetime': digest('fixture-shell:' + str(shell.pid)), 'windowIdentity': digest('private-pty:' + str(shell.pid)),
              'shellIdentity': 'fixture-shell-' + str(shell.pid), 'sessionId': thread, 'directory': project, 'acceptedTurnsHash': baseline,
              'busy': False, 'queueEmpty': True, 'pendingReceiptsResolved': True}
    if value is None:
        if 'CLAWDAD_FIXTURE>' not in visible():
            raise RuntimeError('The original shell is not observed.')
        return {**common, 'kind': 'shell', 'draft': {'text': '', 'hash': digest(''), 'verified': True, 'provenance': 'rendered-composer'}}
    if not status or value['processIdentity'] != status['processIdentity'] or not verify_draft() or 'esc to interrupt' in visible():
        raise RuntimeError('The exact idle status and draft are unavailable.')
    return {**common, **value, 'kind': 'agent', 'authorizationHome': str(base / status['profile']), 'email': status['email'],
            'model': status['model'], 'reasoningEffort': status['reasoningEffort'], 'settingsVerified': True, 'launchPolicyVerified': True,
            'shellWillRemain': True, 'executable': '/opt/homebrew/bin/codex', 'resumeOptions': ['--sandbox', 'read-only', '--ask-for-approval', 'never'],
            'images': [], 'draft': {'text': known, 'hash': digest(known), 'verified': True,
                                  'provenance': 'unchanged-native-paste' if '\n' in known else 'rendered-composer'}}

def launch(profile, request_id):
    global current, known, status
    if profile not in ('cody', 'sun') or owner() is not None:
        raise RuntimeError('Launch requires an empty private shell and an approved retained account.')
    wait(lambda text: 'CLAWDAD_FIXTURE>' in text)
    args = ['/usr/bin/env', 'CODEX_HOME=' + str(base / profile), '/opt/homebrew/bin/codex', 'resume', thread, '--cd', project, '--no-alt-screen',
            '--model', 'gpt-6-astra', '-c', 'model_reasoning_effort="low"', '-c', 'cli_auth_credentials_store="keyring"',
            '-c', 'sqlite_home=' + encoder.encode(str(fixture / 'index')), '--sandbox', 'read-only', '--ask-for-approval', 'never']
    screen.reset()
    shell.send(' '.join(shlex.quote(arg) for arg in args))
    time.sleep(0.25)
    shell.send('\r')
    known = ''
    status = None
    wait(lambda text: 'CLAWDAD_CONTINUITY_9F3A' in text and 'Ask Codex' in text and 'Resuming session' not in text, 35)
    current = profile
    capture_status(profile)
    receipts[request_id] = {'state': 'dispatched', 'effect': 'launch', 'processIdentity': status['processIdentity']}
    return observe()

def insert(text, request_id):
    global known
    before = observe()
    if before['kind'] != 'agent' or known or not text or len(text.encode()) > 16 * 1024 or any(ord(c) < 32 and c not in '\n\t' for c in text):
        raise RuntimeError('Only a bounded synthetic draft into the exact empty fixture composer is allowed.')
    # Bracketed paste is the CLI interface; Enter and Tab are never sent here.
    shell.send('\x1b[200~' + text + '\x1b[201~')
    known = text
    wait(lambda _: verify_draft())
    if owner()['processIdentity'] != before['processIdentity']:
        raise RuntimeError('Fixture input owner changed.')
    receipts[request_id] = {'state': 'dispatched', 'effect': 'draft'}
    return observe()

try:
    wait(lambda text: 'CLAWDAD_FIXTURE>' in text)
    for line in sys.stdin:
        request = json.loads(line)
        serial += 1
        try:
            action = request['method']
            args = request.get('args', {})
            if action == 'launch':
                result = launch(args['profile'], args['requestId'])
            elif action == 'observe':
                receive()
                result = observe()
            elif action == 'insert':
                result = insert(args['text'], args['requestId'])
            elif action == 'stop':
                before = observe()
                if before['kind'] != 'agent' or before['processIdentity'] != args['processIdentity'] or owner()['pid'] != before['pid']:
                    raise RuntimeError('Stop is restricted to the exact idle private fixture owner.')
                # This PID was launched under our one persistent fixture shell,
                # verified foreground, exact history, idle state and unchanged input.
                receipts[args['requestId']] = {'state': 'dispatched', 'effect': 'stop'}
                os.kill(before['pid'], signal.SIGTERM)
                wait(lambda text: owner() is None and 'CLAWDAD_FIXTURE>' in text)
                known = ''
                status = None
                result = observe()
            elif action == 'receipt':
                result = {'requestId': args['requestId'], 'durable': False, **receipts.get(args['requestId'], {'state': 'not_dispatched'})}
            elif action == 'finish':
                result = {'samePTY': tty, 'shellPid': shell.pid, 'acceptedTurnsHash': accepted(), 'baselineHash': baseline, 'modelTurnsSent': 0}
            else:
                raise RuntimeError('Unsupported fixture control.')
            write('rendered-%03d.txt' % serial, visible())
            print(encoder.encode({'id': request['id'], 'result': result}), flush=True)
        except Exception as error:
            write('failure-%03d.txt' % serial, visible())
            print(encoder.encode({'id': request['id'], 'error': str(error)}), flush=True)
finally:
    # Stop only this fixture's own descendants and PTY. A real user PID is never
    # accepted by this driver, and no Terminal application window was created.
    try:
        value = owner()
        if value:
            os.kill(value['pid'], signal.SIGTERM)
            for _ in range(50):
                receive()
                if owner() is None:
                    break
    finally:
        shell.close(force=True)
        write('synthetic-terminal.raw.txt', raw)

File: test/fixtures/codex-account-handoff-pty.mjs
// Live, bounded CLI fixture using the two already approved retained accounts.
// No browser/login, ordinary model turn, real Terminal tab or named snapshot.
import fs from 'node:fs/promises';
import os from 'node:os';
import path from 'node:path';
import {spawn} from 'node:child_process';
import {createInterface} from 'node:readline';
import {createHash} from 'node:crypto';
import {CodexAccountTerminalHandoff} from '../../lib/codex-account-terminal-handoff.mjs';
import {CodexAccountProfileProcess} from '../../lib/codex-account-profile-process.mjs';
import {CodexManagedLogin} from '../../lib/codex-managed-login.mjs';
import {compareAccountRuntimeConfiguration} from '../../lib/codex-account-layout.mjs';
import {researchSave} from '../../lib/research-budget.mjs';

const name=process.argv[2];
if(process.argv[3]!=='--approved-two-account-fixture'||!/^handoff-pty-[a-z0-9-]+$/.test(name||''))throw Error('Use a new handoff-pty-* name and the approved two-account fixture flag.');
const base=path.join(os.homedir(),'Library/Application Support/ClawDad/Accounts/verification-2026-09-15'),fixture=path.join(base,'thread-continuity-1'),root=path.join(fixture,name);
try{await fs.lstat(root);throw Error('Inspect the existing receipt; never replay the fixture name.');}catch(error){if(error.code!=='ENOENT')throw error;}
const candidate=path.resolve('native/macos/dist/candidates/codex-account-switch-2026-09-15');
const python=path.join(candidate,'tui-probe-venv/bin/python');
const accounts={},configurations={};
for(const [profile,email] of [['cody','codyshanemitchell@gmail.com'],['sun','playinthesunwithme@gmail.com']]){
  const connection=new CodexAccountProfileProcess({home:path.join(base,profile),binary:'/opt/homebrew/bin/codex'});
  try{
    await connection.connect();accounts[profile]=await new CodexManagedLogin({rpc:(...args)=>connection.request(...args)}).identity(email);
    configurations[profile]=(await connection.request('config/read',{cwd:path.join(fixture,'project'),includeLayers:false})).config;
  }finally{connection.close();}
}
const configuration=await compareAccountRuntimeConfiguration(configurations.cody,configurations.sun);
if(!configuration.equivalent)throw Object.assign(Error('The two fixture configurations differ; inspect changed fields without altering permissions.'),{changedFields:configuration.changedFields});
const child=spawn(python,['test/fixtures/codex-account-handoff-pty.py',name],{cwd:process.cwd(),stdio:['pipe','pipe','pipe']});
const output=createInterface({input:child.stdout}),pending=new Map();let serial=0,stderr='',exited=false,exitCode;
const exit=new Promise(resolve=>child.on('exit',code=>{exited=true;exitCode=code;for(const item of pending.values()){clearTimeout(item.timer);item.reject(Error('The private fixture process exited.'));}pending.clear();resolve(code);}));
child.stderr.on('data',data=>{stderr=(stderr+data).slice(-4000);});
output.on('line',line=>{
  try{const value=JSON.parse(line),item=pending.get(value.id);if(item){pending.delete(value.id);clearTimeout(item.timer);value.error?item.reject(Error(value.error)):item.resolve(value.result);}}
  catch{child.stdin.end();}
});
const rpc=(method,args={})=>new Promise((resolve,reject)=>{
  if(exited)return reject(Error('Fixture closed'));
  const id=++serial,timer=setTimeout(()=>{pending.delete(id);reject(Error('The bounded fixture action timed out.'));},60000);
  pending.set(id,{resolve,reject,timer});child.stdin.write(JSON.stringify({id,method,args})+'\n');
});
const receipt={version:1,state:'running',modelTurnsSent:0,signInActions:0,realTerminalWindows:0,startedAt:new Date().toISOString(),configuration,
  accounts:Object.fromEntries(Object.entries(accounts).map(([name,value])=>[name,{accountKey:value.accountKey,email:value.subscription.email,remainingPercent:value.remainingPercent,ordinaryUsageAllowed:value.ordinaryUsageAllowed}])),transitions:[]};
const fingerprint=value=>createHash('sha256').update(JSON.stringify(value)).digest('hex');
const attachIdentity=value=>{
  if(value.kind==='agent'){
    const profile=value.authorizationHome===path.join(base,'cody')?'cody':value.authorizationHome===path.join(base,'sun')?'sun':null;
    if(!profile||value.email!==accounts[profile].subscription.email)throw Error('The live local account differs from its exact retained sign-in.');
    return {...value,accountKey:accounts[profile].accountKey,accountVerified:true};
  }return value;
};
const driver={
  observe:async()=>attachIdentity(await rpc('observe')),
  permit:async()=>{}, // This fixture's explicit flag is scoped to its own PTY.
  reconcile:async args=>rpc('receipt',{requestId:args.requestId}),
  verifyOwnership:async({requestId,observed})=>{const value=await rpc('receipt',{requestId});return value.effect==='launch'&&value.processIdentity===observed.processIdentity;},
  stopIdle:async({requestId,source})=>{await rpc('stop',{requestId,processIdentity:source.processIdentity});return {state:'exited'};},
  resumeExact:async({requestId,target})=>{await rpc('launch',{requestId,profile:target.authorizationHome===path.join(base,'sun')?'sun':'cody'});return {state:'started'};},
  restoreDraft:async({requestId,source})=>{await rpc('insert',{requestId,text:source.draft.text});return {state:'inserted'};},
};
try{
  let source=attachIdentity(await rpc('launch',{profile:'cody',requestId:'fixture-start'}));
  const draft='HANDOFF_BEGIN 🌿\n'+('Synthetic unsent research text. Preserve spaces, Unicode Ω and lines.\n').repeat(48)+'HANDOFF_END';
  source=attachIdentity(await rpc('insert',{requestId:'initial-draft',text:draft}));
  const controller=new CodexAccountTerminalHandoff({root:path.join(root,'recovery'),driver});
  await researchSave(path.join(root,'evidence.json'),receipt);
  for(const profile of ['sun','cody']){
    const target={accountKey:accounts[profile].accountKey,authorizationHome:path.join(base,profile),accountVerified:true,configurationVerified:true,
      configurationHash:configuration.sourceHash,layoutFingerprint:fingerprint({sessions:await fs.realpath(path.join(base,profile,'sessions')),index:path.join(fixture,'index'),permissions:'read-only/never'})};
    const args={operationId:'fixture-to-'+profile,requestId:'fixture-click-'+profile,source,target,confirmed:true};
    const result=await controller.run(args);
    const repeated=await new CodexAccountTerminalHandoff({root:path.join(root,'recovery'),driver}).run(args);
    if(result.phase!=='verified'||repeated.phase!=='verified')throw Error('Handoff did not finish with a verified receipt.');
    const current=await driver.observe();
    if(current.sessionId!==source.sessionId||current.tty!==source.tty||current.draft.text!==draft||current.accountKey!==target.accountKey)throw Error('Exact live handoff result did not match.');
    receipt.transitions.push({profile,state:result.phase,sessionId:current.sessionId,tty:current.tty,processId:current.pid,processIdentity:current.processIdentity,
      accountKey:current.accountKey,email:current.email,model:current.model,effort:current.reasoningEffort,draftBytes:Buffer.byteLength(draft),draftHash:current.draft.hash,
      draftVerification:current.draft.provenance,acceptedTurnsHash:current.acceptedTurnsHash,effects:Object.fromEntries(Object.entries(result.effects).map(([key,value])=>[key,{requestId:value.requestId,result:value.result}]))});
    await researchSave(path.join(root,'evidence.json'),receipt);source=current;
  }
  receipt.final=await rpc('finish');receipt.state='verified_same_private_pty_local_resume_and_draft';
}catch(error){receipt.state='needs_attention';receipt.failure=error.message;process.exitCode=1;}
finally{
  child.stdin.end();const timer=setTimeout(()=>child.kill('SIGTERM'),15000);
  await exit;clearTimeout(timer);output.close();receipt.driverExitCode=exitCode;receipt.finishedAt=new Date().toISOString();
  if(stderr){receipt.driverDiagnostic=stderr;receipt.state='needs_attention';process.exitCode=1;}
  await researchSave(path.join(root,'evidence.json'),receipt);
  console.log(JSON.stringify(receipt));
}
